👌 IMPROVE: Ensure len(state.src) == len(state.srcCharCode)

geebos opened the pull request:
```python
def normalize(state: StateCore) -> None:

    # Normalize newlines
    string = NEWLINES_RE.sub("\n", state.src)

    # Replace NULL characters
    string = NULL_RE.sub("\uFFFD", string)

    state.src = string
```

When updating state.src, state.srcCharCode should be updated synchronously.

hukkinj1 commented on markdown_it/rules_core/state_core.py:
```diff
@@ -12,3 +12,7 @@ def __init__(self, src: str, md, env, tokens=None):
         self.tokens: List[Token] = tokens or []
         self.inlineMode = False
+
+    def set_src(self, src):
+        self.src = src
+        self.srcCharCode = [ord(c) for c in src] if src is not None else []
```
```suggestion
    @property
    def src(self):
        return self._src

    @src.setter
    def src(self, value):
        self._src = value
        self.srcCharCode = [ord(c) for c in self.src] if self.src is not None else []
```

hukkinj1 commented on markdown_it/rules_core/normalize.py:
```diff
@@ -17,4 +17,4 @@ def normalize(state: StateCore) -> None:
     string = NULL_RE.sub("\uFFFD", string)
 
-    state.src = string
+    state.set_src(string)
```
```suggestion
    state.src = string
```

hukkinj1 reviewed (commented):
Nice! I made suggestions that use the `@property` decorator to keep API identical to JS markdown-it.

With those changes I believe we can also remove line 10 from the `__init__` that sets `self.srcCharCode`.

hukkinj1 commented:
Also, might make sense to add the `@property` to `StateBase` base class only, to get identical behavior in all `StateBase` subclasses.

geebos commented:
> Also, might make sense to add the `@property` to `StateBase` base class only, to get identical behavior in all `StateBase` subclasses.

Nice suggestions! I have added the `@property` to `StateBase`. In addition, I have added `srcCharCode` to `@property` to avoid assigning srcCharCode from outer codes.

hukkinj1 commented on markdown_it/ruler.py:
```diff
@@ -25,6 +25,19 @@ def __init__(self, src: str, md, env):
+        self._src = value
+        self._srcCharCode = [ord(c) for c in self.src]
+
+    @property
+    def srcCharCode(self):
+        # To avoid assigning srcCharCode
+        return self._srcCharCode
```
```suggestion
```

I think we have to remove this property and keep `srcCharCode` writeable. We write it directly in `StateBlock.__init__`. It seems to me like a performance optimization, but @chrisjsewell knows better.

hukkinj1 reviewed (changes requested):
We should be able to remove the line
```python
self.srcCharCode = [ord(c) for c in src] if src is not None else []
```
in `StateInline.__init__`.

In `StateBlock.__init__` we can make the lines
```python
        self.src = src
        if srcCharCode is not None:
            self.srcCharCode = srcCharCode
        else:
            self.srcCharCode = [ord(c) for c in src] if src is not None else []
```
something like
```python
        if srcCharCode is not None:
            self._src = src
            self.srcCharCode = srcCharCode
        else:
            self.src = src
```

This will keep the (presumably) performance improvement in place.

hukkinj1 commented:
Looking good @geebos ! Did you see this comment https://github.com/executablebooks/markdown-it-py/pull/108#pullrequestreview-558823485 . You think those two changes can be used here too?

geebos commented:
> Looking good @geebos ! Did you see this comment [#108 (review)](https://github.com/executablebooks/markdown-it-py/pull/108#pullrequestreview-558823485) . You think those two changes can be used here too?

Thanks for reminding. I have added a `set_underline_src` method in `StateBase`, because we cant set `_src` directly in `StateBlock`.

hukkinj1 commented:
> we cant set _src directly in StateBlock.

What makes you think so? I believe this should work just fine as long as there's only one preceding underscore (`_src`), not two (`__src`). In case it's mypy that's giving you trouble, you can type annotate `self._src` in `StateBase` after the class declaration like
```python
class StateBase:
    _src: str
```

geebos commented:
You are right. I confuse them.

hukkinj1 commented:
Nice, this looks very good to me!

chrisjsewell commented:
thanks!
